File: src/java/magma/api/collect/Joiner.java
package magma.api.collect;

import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;

public record Joiner(String delimiter) implements Collector<String, Option<String>> {
    public static Joiner empty() {
        return new Joiner("");
    }

    @Override
    public Option<String> createInitial() {
        return new None<String>();
    }

    @Override
    public Option<String> fold(Option<String> maybe, String element) {
        return new Some<String>(maybe.map((String inner) -> {
            return inner + this.delimiter + element;
        }).orElse(element));
    }
}

File: src/node/magma/api/collect/Joiner.ts
import { Option } from "../../../magma/api/option/Option";
import { Collector } from "../../../magma/api/collect/Collector";
import { List } from "../../../magma/api/collect/list/List";
import { None } from "../../../magma/api/option/None";
import { Some } from "../../../magma/api/option/Some";
export class Joiner implements Collector<string, Option<string>> {
	delimiter: string;
	constructor (delimiter: string) {
		this.delimiter = delimiter;
	}
	static empty(): Joiner {
		return new Joiner("");
	}
	static joinOrEmpty(items: List<string>, delimiter: string, prefix: string, suffix: string): string {
		return items.query().collect(new Joiner(delimiter)).map((inner: string) => prefix + inner + suffix).orElse("");
	}
	createInitial(): Option<string> {
		return new None<string>();
	}
	fold(maybe: Option<string>, element: string): Option<string> {
		return new Some<string>(maybe.map((inner: string) => inner + this.delimiter + element).orElse(element));
	}
}
